### DISCUSSION TITLE
Discussion to consider integration of Roblox's recently open sourced Luau

### DISCUSSION THREAD

**Discussion by Reinitialized** [luvit/luvit#1167] [Ideas]
[Roblox](https://roblox.com) has recently opened their heavily modified but Lua 5.1 backwards compatible fork of Lua, called [Luau](https://github.com/roblox/luau). Luau includes a variety of improvements to Lua, from implementing a typing system to drastic improvements in both runtime and compiler performance. Instead of going into all the details here, [I'll just leave the Luau website here](https://luau-lang.org/). I think Luvit would hugely benefit from implementing Luau, making it a potential to become a competitor to NodeJS. Thoughts?

**Comment by truemedian** [comment]
While Luau does introduce improvements to PUC Lua, benchmarks show that it continues to lag behind LuaJIT in interpreter performance (with JIT disabled). So introducing Luau in place of LuaJIT (Luvit's default runtime) would incur only performance losses, would lose out on both JIT compilation and LuaJIT's FFI interface.

Another glaring issue with Luau is that the Lua api that it exports does not use the C calling convention, rather it uses C++'s compiler-specific, name-mangling calling convention. This means that Luvi (the native portion of Luvit), along with all of the libraries it includes (such as lua-openssl, lua-compat5.3, lrexlib, lpeg and others) simply cannot use it, as they does not know or understand how to call the runtime in order to do anything.

**Comment by rdw-software** [comment]
> I think Luvit would hugely benefit from implementing Luau, making it a potential to become a competitor to NodeJS

How would it benefit, and how would using Luau instead of LuaJIT enable any Lua runtime to become a competitor to NodeJS? Always assuming that this is actually a worthwhile or even realistic goal, and that there are enough people available to realize it.

I'm unclear on this part, since the biggest selling point of LuaJIT is speed (and a smaller one its FFI). The biggest selling point of NodeJS isn't *just* speed, it's also the large JS ecosystem and forced adaption of the language through browser vendors. Luau seems to offer no upgrades over LuaJIT (and Luvit) in any of these areas, so forgive me if I fail to see what integrating it would actually accomplish.

**Comment by Bilal2453** [comment]
This is a weird suggestion no doubt. Assuming Luau had any benefits to its side (to my knowledge none), that would still require a whole rewrite of the whole Luvit ecosystem; each. single. one. That is not only a huge work to do, even bigger than the Luvit 3 rewrite, but it also sounds pointless sadly.

On the other hand, if the goal is to make Luvit a real competitor to Node.js, the right decision would be to rewrite the Luvit API to make it more compiling, better docs, bigger community targeting, etc; luckily, that's indeed the goal with the whole Luvit 3 rewrite. See Issue #1135.

Edit: Also do note, Luau most compiling selling point is its sandbox, something Luvit doesn't really need at all.

**Comment by LeXinshou** [comment]
I wasn't logged in to the github for some reason but when I saw this issue, I couldn't stop myself logged-in again to make a comment here.

It seems like you do love Luau, no doubt there, we all love Lua in general anyway. Luau has different setup, it's true it has type system and some other cool feature, I must admit that's a good thing /BUT/ if you compare Luvit with Luau, you are out of your mind and your opinion seems so poor. Luvit in general doesn't need to be a competitor, every runtime has it's own benefit, why even bother with nodeJS? NodeJS itself has JIT functionalities, so do Luvit. Luau doesn't have the JIT yet, tho they are doing pretty impressive things but if we care about luvit's ecosystem, it's good how it is now.

Luau is no where near to Luvit or there nothing Luvit can be benefited from Luau at all.

**Comment by Reinitialized** [comment]
I am now aware of the rework required from how Luau exposes its API (only compatible with C++), which in itself is a major roadblock. Maybe using the words "competitor" to NodeJS was a bad idea, but what I mean by this is in terms of having more options to build a project of choice without feeling locked into one ecosystem due to it's overwhelming dominance (making a Discord bot for example. I made one in Luvit in the past, and run into all sorts of issues due to lack of dependencies and weird little unexpected bugs, ultimately converting it to NodeJS due to it's overall stability and support).

I am also aware of the performance difference, which although seems *relatively* small to me, still does make an impact. In regards to Luau not being JIT, the reasoning behind this was due to iOS, as Roblox considered a JIT implementation in the past, but didn't want to make a split in performance differences between iOS and all other platforms which allowed JIT. With Luau now being open sourced, this may get changed either by Roblox or through a fork, but at the moment is an understandable concern.

The biggest reasoning I'm excited about Luau being open source is the amount of time invested into bringing both (from my understanding) a compiler and runtime Typed Lua, something not even TypeScript offers. This being the initial open source release of Luau, and being backed by a well funded company whose entire development API for end users is itself Lua, I imagine it'll only get better with time. My intent with this issue is to merely bring discussion to the table and see concerns such as the ones already brought up, along with potential interest as of this initial release.

**Comment by dlannan** [comment]
Just want to add in some notes about luvit moving forward. People keep referring to Luajit as dead - this is not true, the company I work for (fmad.io) recently engaged Mike P to add a bunch of awesome features to Luajit (interning strings and string buffers). We will be doing more, and its all going out to the main branch. Having used Lua for 20 years, and Luajit since its inception, I would _highly_ recommend staying on the Luajit path. There are _alot_ of updates coming, and its performance is lightyears above anything you will get from PUC. And most importantly, ffi - this is such a major beneficial feature its very hard to explain to people just how powerful it makes luajit. So, from my perspective, I would very much dissuade people from moving from luajit to Luau. The benefits are just not there, considering the amount of effort needed and time taken to make it happen. 

I also believe that languages DO NOT have to keep changing and having features added - luajit is a brilliant example of this. We spent around a year assessing alternatives to Luajit before getting MP on board. Im so glad we did, every other solution was going to be _years_ to change over. Luajit + FFI + C.. has a _deep_ archive of libraries and capability. And it is highly flexible. With the tools around luvit, luvi and luv you have a potential ecosystem that could easily be one of the most powerful hidden gems around.

On the above suggestions. I think follow Tim's guidance here - extract some of the libraries (openssl etc) and consolidate and modernize what is available. I personally (my own company) have delivered a few solutions using luvit (albeit heavily miodified) for a number of different types of uses: web server for an entertainment management site, a simulation system with luvit as a backend manager for websockets and sim module intergrations, and I also made a little GUI attached to luvit using a couple of different systems. Its quite diverse and powerful.

I would be interested in contributing, if there were clear goals forward for this. I would probably drop use of luvit if it went down the PUC route.

**Reply by SinisterRectus** [reply]
Thanks for the feedback. Is any of this work available for viewing?

**Comment by dlannan-fmad** [comment]
Yes. Some of the prototype work and the simulation system is - but luvit is only a backend so there isn't much to see. It is not open source (yet). It will be early next year. 
The front end uses apache or nginx (I did have a working luvit, but there were some problems with Grav). 
https://seer.kakutai.com:2727      login: seerguest  pw SeerGuest2020
And https://seer.kakutai.com:2727/scenarist?projectid=27&sceneid=84

This site was originally all done with luvit - in less than 2 months (It has been converted to apache in recent times).
https://www.outofthesquare.net.au/
Includes a full backend editing system for the content authors. I have attached the original project here if it interests you :) 
https://drive.google.com/drive/folders/1Q-twLp-zeMOZA0Eb1BBm2INgsl3JZ8Wm?usp=sharing

I have a few open source luvit samples that were used for testing a number of methods used in these apps:
https://github.com/dlannan/luvit-bullet3  (windows only)
https://github.com/dlannan/luvit-newton (windows only)
https://github.com/dlannan/luvit-bgfx (windows only)
Also worked with a company called TITAN IM (  http://titanim.net/ ) to build a DIS integration that used a little luvit server. 

Most of the above is 2-3 years or more old. A current system I am working on with luvit is a game server here:
https://github.com/dlannan/swampy
The demo server (wip) is here: 
swampy.kakutai.com:5000   login: admin@test.com   pw: admin
(Note, Im working on getting the luvit websockets to support fragmentation).

Supports modular game networking for the Defold game engine. 
Ive used luvit/luajit in a number of tools and we use luajit exclusively for our complete capture systems at fmad.io. We have our own network stack and system for luajit that is bound to FPGA, otherwise luvit may have been an option :)

Hope this is the sort of info you were wanting. Feel free to ask more on Discord if you need:  dlannan#1808

**Comment by creationix** [comment]
I finally dug into this and I really like the design of luau.  I agree it might be slightly slower than luajit, but the other benefits make it worth it IMHO.

The main issue is the language is different enough that most the luvit ecosystem will have issues.  It's bad enough trying to support luajit and lua, luau is a different language as well.

I would love to do a spike and see how fast I can make something that luau + network primitives and start from scratch on the userspace library side.

We can keep the luvi + zip design since it appears luau is embeddable just like luajit.

There is no fast FFI for calling C functions (and honestly we barely use it in luvit anyway because libuv is callback based).  So I guess the best thing is more of the same with hand-written host bindings in C?

**Reply by Bilal2453** [reply]
Personally I prefer something closer to the [Pluto language](https://pluto-lang.org) (although in its current form it is its own runtime with its own networking etc), it seems much more mature to me design wise minus the typing and it is bytecode backwards compatible with PUC Lua 5.4, or even something that compiles to Lua such as Teal.  This way it is possible to support multiple languages as long as they eventually compile to Lua/are Lua compatible.   We can see this in the JS ecosystem for example where the runtime would just compile TS/CoffeeScript/etc to JS and hope for the best, basically think of Luajit as being the V8 engine. I believe it is possible to compile Luau to PUC Lua as well.

On another note, [Lune](https://github.com/lune-org/lune) is a new competitor to Luvit, built with Luau and aims to be the Deno of Lua (contrary to Luvit being the Nodejs of Lua).
At this point we just need the Bun of Lua with a Zig backend 😝

**Comment by SinisterRectus** [comment]
> On another note, [Lune](https://github.com/lune-org/lune) is a new competitor to Luvit, built with Luau and aims to be the Deno of Lua (contrary to Luvit being the Nodejs of Lua).

So, closed as completed?

**Reply by Bilal2453** [reply]
Perhaps, but this is about Luvit-support specifically.

This should probably become part of the Luvit 3.0 discussion though, if not its own discussion instead of being an issue, the topic should instead be "Native support for other Lua dialects" instead of being Luau specific, I do think having such support in the future is a good idea, and Luau is definitely one of the possible options for that as well as things like Teal.

**Reply by preindex** [reply]
Yeah, I do think that considering the integration of other Lua dialects would be a worthwhile endeavor. This is keeping in mind that LuaU has evolved and matured in the last 4 years since this original posting of this discussion.

As someone who has been used to developing with LuaU due to ROBLOX, the libraries that Lune provides are extremely helpful to me, and provide a sense of familiarity and accessibility. That said, I've also used Luvit as it is now, and while it works well, it just feels off, especially if you consider my ROBLOX background. That's why I 100% back these integrations.

My support also comes from the concern of fragmentation. I don't want to have multiple package managers and runtimes for what is essentially a dialect of the same language. Especially for such a niche benefit. I've experienced this kind of thing with node, bun, and deno. It's not fun to deal with, and I'd hate to see Lua go down the same path.

Lune wants to take LuaU beyond just ROBLOX development, and while I respect that, Luvit does the same, but for Lua. The only difference is that they're different dialects. I don't want to use Lune just because I like the way it functions, and then use Luvit because I want access to a larger library. It's impractical and inefficient.

While I'll use Lune for now as it's more tailored for my needs, I'd much rather see Luvit evolve into an "all-in-one solution" that I can use for anything Lua-related, which would be far more convenient, especially since I'm already familiar with it.

**Reply by truemedian** [reply]
> My support also comes from the concern of fragmentation.

Unfortunately Luau has adopted a model where they force themselves into fragmentation, the Luau headers are not ABI compatible with any existing version of Lua and Luau does not provide *any* form of external binary library loading (which forms a large majority of meaningful libraries). The combination of both essentially requires the Luau environment to adopt their own package environment because the existing Lua ecosystem cannot work with Luau.

It's worth noting that adopting Luau into Luvit requires adopting Luau into Luv, which has almost no incentive to support the project since its primary method of distribution is as an external binary library.

**Reply by Bilal2453** [reply]
> the integration of other Lua dialects would be a worthwhile endeavor

How so? The only reason I may like support for a different dialect is frankly types, it just makes documenting the code / auto-generating code editor suggestions a bit easier. Beside that I can't see how it is of any value frankly. You can use Luau with Luvit right now as far as I am concerned by compiling your Luau code into PUC Lua code (there exists multiple tools for that).
My main concern with Luau is frankly the fragmentation issue it made worse, when it could have taken an approach like Teal's which ensures high compatibility.

The familiarity argument is not that great either, what is familiar to you is not necessarily familiar to the rest of community, the Luau approach to a lot of things are frankly far from that of the rest of the Lua community.  Since this is highly subjective it is hard to decide in favor of it.  

You mentioned JS but the situation in the JS ecosystem is far better, most runtimes has Nodejs API support or are working on it (Deno 2 improved their compatibility and Bun are working on it), especially with the existence of Babel, and almost all of the JS package managers are compatible with NPM, but the fragmentation in question here is deeper, for JS it is just [ECMAScript](https://en.wikipedia.org/wiki/ECMAScript) and its versions (which you can go forward/backward using Babel, and TS just compiles to ECMAScript at the end), for Lua it is so many more things, for example in Luau you get a lot of the API completely removed making it impossible to integrate with normal Lua code. In other runtimes of Lua you also get a lot of things stripped/added, this happens because of the nature of the language where it encourages the host to do its own integration (kind of the whole purpose of the language).
Teal compiler does do something similar to Babel, you can write Teal code, and compile to any PUC Lua version from 5.1 to 5.4, would be neat if we also had this but also for other dialects, although that is easier said than done considering outside of PUC Lua things get out of hand real fast, e.x. Luau removing most of the Lua standard library, adding new metamethods, altering the behavior of globals and environments, altering the behavior of other standard functions, etc. See [Luau Compatibility](https://luau.org/compatibility) page.

The fragmentation issue is really sad, but it is also built-into the language, it is what for the language is made. I lost all hope of fixing the situation when Luau was released as open-source knowing that people will start using it outside of Roblox, further breaking the compatibility (and in this case it is even worse because it completely breaks ABI/bytecode).

We should not focus on runtime/language compatibility like JS is doing because frankly that is of no value, instead the focus shall be building a great runtime for the user in terms of standard library support and developer experience, and minor nice-to-have things like support for typed Lua subsets may come at a later stage. At the end of the day, languages and runtimes are tools, and the user will choose to use the most suited, capable and easy-to-use tool (and right now, that is not necessarily Luvit).

**Reply by preindex** [reply]
You guys raise some pretty important points. Yeah, LuaU is very, very different from just Lua, and familiarity is subjective. Plus, it does also obscure a bunch of things in the name of security (which doesn't really make sense since our use-case is different). 

I guess my support was misguided, and wasn't directly because of the language, but the built-in libraries that it provided, which wouldn't be a great argument. I'd still stand by supporting things like compound operators and, of course, type annotations.
